src/analyses/threadEscape.ml

```diff
@@ -42,76 +44,124 @@ struct
       if M.tracing then M.tracel "escape" "mpt %a: %a\n" d_exp e Queries.LS.pretty a;
       D.empty ()
 
+  let thread_id ctx =
+    ctx.ask Queries.CurrentThreadId
+
+  (** Emit an escape event:
+      Only necessary when code has ever been multithreaded,
+      or when about to go multithreaded. *)
+  let emit_escape_event ctx escaped =
+    (* avoid emitting unnecessary event *)
+    if not (D.is_empty escaped) then
+      ctx.emit (Events.Escape escaped)
+
+  (** Side effect escapes: In contrast to the emitting the event, side-effecting is
+      necessary in single threaded mode, since we rely on escape status in Base
+      for passing locals reachable via globals *)
+  let side_effect_escape ctx escaped threadid =
+    let threadid = G.singleton threadid in
+    D.iter (fun v ->
+        ctx.sideg v threadid) escaped
+
   (* queries *)
   let query ctx (type a) (q: a Queries.t): a Queries.result =
     match q with
-    | Queries.MayEscape v -> D.mem v ctx.local
+    | Queries.MayEscape v ->
+      let threads = ctx.global v in
+      if ThreadIdSet.is_empty threads then
+        false
+      else begin
+        let possibly_started current = function
+          | `Lifted tid ->
+            let threads = ctx.ask Queries.CreatedThreads in
+            let not_started = MHP.definitely_not_started (current, threads) tid in
+            let possibly_started = not not_started in
+            possibly_started
+          | `Top -> true
+          | `Bot -> false
+        in
+        let equal_current_not_unique current = function
+          | `Lifted tid ->
+            ThreadId.Thread.equal current tid && not (ThreadId.Thread.is_unique current)
+          | `Top -> true
+          | `Bot -> false
+        in
+        match ctx.ask Queries.CurrentThreadId with
+        | `Lifted current ->
+          let possibly_started = ThreadIdSet.exists (possibly_started current) threads in
+          if possibly_started then
+            true
+          else if ThreadIdSet.exists (equal_current_not_unique current) threads then
+            (* Another instance of the non-unqiue current thread may have escaped the variable *)
+            true
```

Comment:
Maybe first check current thread for uniqueness?

Reply:
Good idea.